### PR TITLE
fix: #7

### LINKED ISSUE #7: VM not usable when keep_registered option is used (Hyper-V)
_This issue was originally opened by @godambrosio as hashicorp/packer#10332. It was migrated here as a result of the [Packer plugin split](https://github.com/hashicorp/packer/issues/8610#issuecomment-770034737). The original body of the issue is below._

<hr>

When filing a bug, please include the following headings if possible. Any
example text in this template can be deleted.

#### Overview of the Issue

The VM that keep registered on Hyper-V is not usable because the used disk doesn't exist.

#### Reproduction Steps

1- Build a Hyper-V vm using the option "keep_registered":  "true"
2- Try to start the resulting VM 

### Packer version

Packer v1.6.4

### Simplified Packer Buildfile

{
    "variables":  {
                      "disk_size":  "61440",
                      "iso_checksum":  "CHECKSUM_VALUE",
                      "iso_url":  "URL_TO_WINDOWS_ISO",
                      "restart_timeout":  "2m"
                  },
    "builders":  [
                     {
                         "boot_command":  [
                                              "\u003center\u003e\u003cwait\u003e\u003center\u003e\u003cwait\u003e\u003center\u003e"
                                          ],
                         "boot_wait":  "1s",
                         "communicator":  "winrm",
                         "cpus":  2,
                         "disk_size":  "{{user `disk_size`}}",
                         "enable_secure_boot":  true,
                         "enable_virtualization_extensions":  true,
                         "first_boot_device":  "DVD",
                         "generation":  2,
                         "iso_checksum":  "md5:{{user `iso_checksum`}}",
                         "iso_url":  "{{user `iso_url`}}",
                         "memory":  4096,
                         "temp_path":  "./tmp",
                         "secondary_iso_images":  [
                                                      "./iso/windows_10_unattend.iso"
                                                  ],
                         "switch_name":  "Default Switch",
                         "type":  "hyperv-iso",
                         "headless":  "true",
                         "keep_registered":  "true",
                         "skip_export" : "true"
                     }
                 ],
   
}

### Operating system and Environment details

OS: windows 10 Professionnel
Hyper-V version : 10.0.19041.1 

### Log Fragments and crash.log files

hyperv-iso: output will be in this color.

==> hyperv-iso: Creating build directory...
==> hyperv-iso: Retrieving ISO
==> hyperv-iso: Trying http://XXXXXXX
==> hyperv-iso: Trying http://XXXXXXX
==> hyperv-iso: http://XXXXXX => D:\XXXXXX
==> hyperv-iso: Creating switch 'Default Switch' if required...
==> hyperv-iso:     switch 'Default Switch' already exists. Will not delete on cleanup...
==> hyperv-iso: Creating virtual machine...
==> hyperv-iso: Enabling Integration Service...
==> hyperv-iso: Mounting os dvd drive D:\XXXX
==> hyperv-iso: Skipping mounting Integration Services Setup Disk...
==> hyperv-iso: Mounting secondary DVD images...
==> hyperv-iso: Mounting secondary dvd drive ./iso/windows_10_unattend.iso ...
==> hyperv-iso: Configuring vlan...
==> hyperv-iso: Setting boot device to "DVD"
==> hyperv-iso: Determine Host IP for HyperV machine...
==> hyperv-iso: Host IP for the HyperV machine: 192.168.64.1
==> hyperv-iso: Starting the virtual machine...
==> hyperv-iso: Waiting 1s for boot...
==> hyperv-iso: Typing the boot command...
==> hyperv-iso: Waiting for WinRM to become available...
    hyperv-iso: WinRM connected.
==> hyperv-iso: #< CLIXML
==> hyperv-iso: <Objs Version="1.1.0.1" xmlns="http://schemas.microsoft.com/powershell/2004/04"><Obj S="progress" RefId="0"><TN RefId="0"><T>System.Management.Automation.PSCustomObject</T><T>System.Object</T></TN><MS><I64 N="SourceId">1</I64><PR N="Record"><AV>Préparation des modules à la première utilisation.</AV><AI>0</AI><Nil /><PI>-1</PI><PC>-1</PC><T>Completed</T><SR>-1</SR><SD> </SD></PR></MS></Obj><Obj S="progress" RefId="1"><TNRef RefId="0" /><MS><I64 N="SourceId">1</I64><PR N="Record"><AV>Préparation des modules à la première utilisation.</AV><AI>0</AI><Nil /><PI>-1</PI><PC>-1</PC><T>Completed</T><SR>-1</SR><SD> </SD></PR></MS></Obj></Objs>
==> hyperv-iso: Connected to WinRM!

### PR DESCRIPTION
**Pull request by remi-espie**
When the `keep_registered` flag was true, the VM stayed in HyperV manager, but its files were deleted anyway, making it impossible to use the VM as-is.

This PR makes the VM's files cleanup look for the `keep_registered` flag, and if set to true, will not delete the VM files, keeping it usable.

However, the VM files stays in the temp folder. We may want to move the files to another location beforehand if the flag is up.

This PR closes #7 .

### CODE REVIEW

File: builder/hyperv/common/step_create_build_dir.go [hunk 1]
```diff
@@ -62,6 +65,12 @@ func (s *StepCreateBuildDir) Cleanup(state multistep.StateBag) {
 	}
 
 	ui := state.Get("ui").(packersdk.Ui)
+
+	if s.KeepRegistered {
```
Comment (lbajolet-hashicorp) [inline builder/hyperv/common/step_create_build_dir.go:69]: I would suggest checking if the build succeeded before keeping the artifacts, we should probably remove it if the build failed or was aborted for any reason.

```suggestion
	_, cancelled := state.GetOk(multistep.StateCancelled)
	_, halted := state.GetOk(multistep.StateHalted)

	if s.KeepRegistered && !(cancelled || halted) {
```

**Review by lbajolet-hashicorp — COMMENTED** [review]
Hey @remi-espie,

Thanks for this PR!

By curiosity, have you tested it in your setup since you hit the problem described in the issue? Did that fix it?

I'm not a Hyper-V expert at all, so I'll have to rely on your expertise here, but the code looks good to me, just want to confirm it does work at least in your env, and I left a quick suggestion regarding the removal of the image.

Let me know when you have time to address the suggestion I left, and I'll do a second pass after that.

**Comment by remi-espie** [comment]
Hello !

Thanks for the great suggestion, I just added it !

To answer your initial question :
> By curiosity, have you tested it in your setup since you hit the problem described in the issue? Did that fix it?
> 
> I'm not a Hyper-V expert at all, so I'll have to rely on your expertise here, but the code looks good to me, just want to confirm it does work at least in your env, and I left a quick suggestion regarding the removal of the image.

I'm not a Hyper-V expert either, I just worked with it recently. I tested it one time and it worked flawlessly: the VM is still registered in Hyper-V and you can use it without issue. However, as I said, the artifacts stay located in the user's TEMP folder, which does not seems to be ideal.

**Comment by lbajolet-hashicorp** [comment]
This is a good point, I'm not sure what the artifact produced by Packer looks like for Hyper-V, but in this case it might be a good idea to export the files in a separate location then?
I don't know how temp files are managed on Windows, but if it's in any way similar to Linux, they might be memory-mapped, and flushed on shutdown, so in this case we lose them on each power-cycle.

This may signal we should have an `output_directory` (through this being Windows, `output_folder` may be more idiomatic) option for the template? This would supersede `keep_registered` in this case, and export the files in a persistent directory.
This option may even be the actual working directory rather than a temporary one, which will not be cleaned-up, unless there's a problem with the build as pointed out in my last review.

If you agree and decide to go with the `output_directory` route, feel free to take inspiration from the [`qemu` plugin for example](https://github.com/hashicorp/packer-plugin-qemu/blob/main/builder/qemu/config.go#L396). Please note that the option is not mandatory here as we pick a default, but there always is a persistent directory for the output, and must not exist before running the build.

Let us know what you think, and thanks for the quick turnaround time!

**Comment by remi-espie** [comment]
For Hyper-V, the artifacts consists of a few folder with data about the VM (like network interface, CPU number, etc) in .vmxc files and the VM as a .vhdx file.

I believe Windows does not flush the TEMP folder at each shutdown, however I also think that an `output_directory` option would be a great addition.

I won't be able to work on this new option in the following weeks, so it may be better to create a new PR that would add the `output_directory` option. What do you think ?

**Comment by lbajolet-hashicorp** [comment]
I don't think we'll be able to work on this soon as well, so regarding this PR that's up to you, I don't mind leaving it open with the discussion readily available, and either you can pick it up when you have time, or someone else may piggyback off it and add the missing attribute.

If you prefer to start with a clean slate, feel free to close it and we'll review the output directory PR when available for sure.

Thanks!

**Comment by remi-espie** [comment]
Hello @lbajolet-hashicorp . I just noticed that the Hyper-V plugin has a `temp_path` parameter which default to `%TEMP%`. By changing it, you can have the VM still registered and working in the directory you want. 
Also, there is already a `output_directory` parameter. It is not used when the `keep_registered` parameter is true though, as it only works for the VM exportation.

What do you think I should do ? I may supersede `temp_path` with `output_directory` when  `keep_registered` is set to true, however it might break backward compatibility. Else, I can go with `temp_path` and call it a day.

**Comment by lbajolet-hashicorp** [comment]
Hi @remi-espie,

Regarding the `temp_path` attribute, if it already exists and makes it possible for users to have their files in the desired location, we can use it in conjunction with `keep_registered` then.
Let's continue with your PR, I would probably suggest adding something to documentation of the attribute so it is stated that the VM's files will reside in temp by default, unless the `temp_path` attribute is set to something else.

We could probably rename the attribute to something more explicit regarding how it's meant to be used (I'm thinking something like `vm_dir`, but others would work), but as you point out, doing so will break compatibility, and I don't think we need to do it now.

In the end then yes, let's continue with this PR so the directory doesn't get wiped, and let's point out that `temp_path` should be set for such cases in the docs, then we should be good to go.

Thanks!

**Review by lbajolet-hashicorp — APPROVED** [review]
LGTM, thanks for the rerolls @remi-espie !

**Comment by remi-espie** [comment]
Hello @pmcevoy, I've tested your configuration, and turns out, `skip_export` mess with the `keep_registered` parameter. I may make both parameter incompatible if used together, but for now, you can use the VMs in their temp path if you remove the `skip_export` parameter (which, is this case, does nothing else than moving the .vhdx file).

**Pull request merged**